File: leetcode/0807_max_increase_to_keep_city_skyline.kt
package leetcode

class Solution {
    fun maxIncreaseKeepingSkyline(grid: Array<IntArray>): Int {
        val row_max = IntArray(grid.size, { _ -> 0 })
        val col_max = IntArray(grid[0].size, { _ -> 0 })

        for (i in 0..grid.size - 1) {
            for (j in 0..grid[0].size - 1) {
                row_max[i] = maxOf(row_max[i], grid[i][j])
                col_max[j] = maxOf(col_max[j], grid[i][j])
            }
        }

        var result = 0

        for (i in 0..grid.size - 1) {
            for (j in 0..grid[0].size - 1) {
                val local_max = minOf(row_max[i], col_max[j])
                result += local_max - grid[i][j]
            }
        }

        return result
    }
}

fun main(args: Array<String>) {
    val solution = Solution()

    val grid = arrayOf(
        intArrayOf(3, 0, 8, 4),
        intArrayOf(2, 4, 5, 7),
        intArrayOf(9, 2, 6, 3),
        intArrayOf(0, 3, 1, 0)
    )

    assert(35 == solution.maxIncreaseKeepingSkyline(grid))
}

File: leetcode/0807_max_increase_to_keep_city_skyline.cpp
// Copyright (c) 2019 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        vector<int> rowMax(grid.size(), 0);
        vector<int> colMax(grid[0].size(), 0);

        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                rowMax[i] = max(rowMax[i], grid[i][j]);
                colMax[j] = max(colMax[j], grid[i][j]);
            }
        }

        int result = 0;

        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                int localMax = min(rowMax[i], colMax[j]);
                result += localMax - grid[i][j];
            }
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<vector<int>> grid = {
        {3, 0, 8, 4},
        {2, 4, 5, 7},
        {9, 2, 6, 3},
        {0, 3, 1, 0},
    };

    assert(35 == solution.maxIncreaseKeepingSkyline(grid));
}
